src/debug.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Debug-logging functionality.
+//!
+//! This module is still present when the `log` feature is disabled, but all
+//! logging operations are redacted. Redaction completely compiles out log
+//! statements: not even the format strings remain in the final binary.
+//!
+//! Manticore code *should not* call into the [`log`] crate directly outside of
+//! this module.
+
+// TODO: Remove this once we start using these macros in Manticore.
+#![allow(unused)]
+
+use core::fmt;
+
+#[cfg(doc)]
+use __raw_log as log;
+
+/// A wrapped Manticore error.
+///
+/// This type should always be referred to as `manticore::Error`. It represents
+/// an error with extra (potentially redacted) information attached. This type
+/// cannot be directly created by users of the library.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Error<E> {
+    inner: E,
+}
+
+impl<E> Error<E> {
+    /// Creates a new `Error`. This function is an implementation detail,
+    /// and should not be called by users.
+    #[doc(hidden)]
+    pub fn __new(inner: E) -> Self {
+        Self { inner }
+    }
+
+    /// Transforms the wrapper error by way of an [`Into`] conversion.
+    ///
+    /// Generally, this function should not be necessary, because Manticore-
+    /// defined error types manually implement the relevant [`From`]
+    /// implementations for `manticore::Error`, which in turn call `cast()`.
+    pub fn cast<F>(self) -> Error<F>
+    where
+        F: From<E>,
+    {
+        Error {
+            inner: self.inner.into(),
+        }
+    }
+
+    /// Gets the wrapper error.
+    pub fn into_inner(self) -> E {
+        self.inner
+    }
+}
+
+impl<E> AsRef<E> for Error<E> {
+    fn as_ref(&self) -> &E {
+        &self.inner
+    }
+}
+
+impl<E> AsMut<E> for Error<E> {
+    fn as_mut(&mut self) -> &mut E {
+        &mut self.inner
+    }
+}
+
+#[cfg(feature = "log")]
```

Comment:
Hmm why is this conditioned on "log"? This might be my doing but I think we can drop it.

Reply:
That was my doing. My thought was to further reduce the number of compiled-in strings, though this has way less impact than the log macros. Moving the gate inside the function implementation rather than gating the entire trait implementation would probably be better. Happy to leave it out entirely if it's over-optimization at this point.

Reply:
Removed